Code Refactoring for Enhanced Performance and Memory Leak Elimination

waltim opened the pull request:
- Refactored the RepositoryWalker class to eliminate memory leaks and improve performance.
- Replaced the use of synchronized with an ExecutorService for more efficient thread management.
- Reused the same thread pool for multiple metric collection tasks.
- General improvements in code organization and exception handling.
- Optimized parallel metric collection to accelerate the revision analysis process.
- Added explicit disposal of unused objects to free up memory as needed.
- Other minor code improvements and fixes to enhance efficiency and robustness.

These changes significantly enhance the performance and stability of Git project analysis.

rafaelcn committed:
Fix typo

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Update vulnerable dependency

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add dependencies

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add a rule for iml files (intellij related)

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add a rule for the idea directory (intellij)

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add basic structure of the walker

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add ANTLR4 generated files

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Update CLI parser

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add a new project flag

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add basic logic for filesystem traversal

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add a new dependency

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add an interval class and refactor package

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Refactor the name of a variable

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add the initial implementation of a repo walker

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Fix code that picks up the extension of the file

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Improve file extension detection

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Make exception class static

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Remove new line from function

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Simplify identification of file extension

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add initial implementation of the metrics collect/

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Remove unused imports

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add log4j configuration file

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add a required dependency

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Improve interval classs

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add implementation of the walker task

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Remove newline

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add list of features and its respective docs

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Change default initial date to 2014

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Update information

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Change indentation and format of file

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add rule for .antlr

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add more visitors for specific JS things

Implements some new visitors such as for class, generators and for let
expressions.

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Mark the option rest parameters

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add an example for a rest parameters

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add new example instances on modules

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add visitors to other statements/declarations

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Fix walker errors

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add other unit tests

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

waltim committed:
implemented methods and tests to collect the use of Spreads, Destructurings, Default Parameters, and Promises.

waltim committed:
Add a metric package

Signed-off-by: waltim <waltimlmm@gmail.com>

waltim committed:
Add a rule for vscode folder

Signed-off-by: waltim <waltimlmm@gmail.com>

waltim committed:
Add support for report writing

Signed-off-by: waltim <waltimlmm@gmail.com>

rafaelcn committed:
Merge pull request #2 from Walter

Finish the implementation of feature for the study

rafaelcn committed:
Replace import var

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Simplify code and refactor function

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Update metric to have a generic value type

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Append the comma as the last argument

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Update code to have the project folder name

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add project, date and commit hash to the summary

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Remove unused stuff and improve the commit metric

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Explicitly flush before closing the report file

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add a static method for the header of a report

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Remove old header method (not used)

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Implement a writer for a results.csv file

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add a file column and remove duplicated data

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Modify the argument declaration

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add metrics for statements and remove function

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add a test for statements

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Move the initial date six months back

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Compare the project name with the last path part

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add a synchronized block and add a branch conditi/

If we can't find the main branch of a repository we should check for
that. This commit adds this condition.

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Use the HEAD if no default branch was found

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Modify build step so it generates a correct JAR

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add a configuration to not generate reduced pom

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

waltim committed:
fix issues to find HEAD branch

rafaelcn committed:
Refactor code and reorder log message

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Replace explicit types

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add type notation to help the inference engine

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Replaace even more explicit types

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Format code

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

waltim committed:
list of all projects and 100 more stargarized in csv, scripts python to clone repositories and running analysis using jar file.

rafaelcn committed:
Move all scripts and CSVs to a folder

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

waltim committed:
change separator to comma and improviments in scripts

waltim committed:
Merge branch 'feature/implement-git-traversal-structure' of github.com:PAMunb/JSMiner into feature/implement-git-traversal-structure

rafaelcn committed:
Add an error map to the summary

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

waltim committed:
python requirements for scripts

waltim committed:
Merge branch 'feature/implement-git-traversal-structure' of github.com:PAMunb/JSMiner into feature/implement-git-traversal-structure

waltim committed:
update .gitignore

rafaelcn committed:
Implement error writer

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Change extension of the error file

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add a general error count at each revision

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Set initial date to be in 2012

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add a new profiler class

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Implement concurrency when parsing/visiting files

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Use integers so we can free the visitor after exec

The atomic reference is still being maintained (or copied) over the
metric value but this is not needed so we can just use the value of the
atomic reference.

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

waltim committed:
update runner and RW to reset before checkout

rafaelcn committed:
Add explicit call to GC after collection

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add a newline

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Improve timer and add some profiler log messages

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Remove unnecessary file

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Update documentation

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add an additional filter for lib/vendor folders

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add a new argument to specify threads on revision

This commit augments the capability of the parser process by giving the
user the ability to specify the number of threads (how many visitors)
working concurrently on a given revision.

Use with care, because this will consume a lot of RAM.

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Improve exception handling and file closing

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Add a revision filter when filtering commits

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Rename variable for clarity

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Improve exception handling

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Modify filter to use a method

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Transform static method into a class

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Rearrange code

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Improve definition of rule

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Modify collect metrics flow

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Fix walker task not having the right number of threads

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Improve exception description

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

rafaelcn committed:
Merge pull request #3 from PAMunb/feature/use-threads-to-collect

Add threads when collecting data from the parser

rafaelcn committed:
Add a test for a rule

Signed-off-by: Rafael Campos Nunes <rafaelnunes@engineer.com>

waltim committed:
add new folder to DirectoriesRules, improve git-log commit list filters and close git before return summaries if mainBranch cannot be found

waltim committed:
close walker after collect all files and remove throwing exceptions to cannot break the execution after catch exception

waltim committed:
Refactored classes for improved memory management and performance.

rafaelcn commented:
Can you open this PR to the branch [feature/implement-git-traversal-structure](https://github.com/PAMunb/JSMiner/tree/feature/implement-git-traversal-structure)?

waltim committed:
- Refactored the RepositoryWalker class to eliminate memory leaks and improve performance.
- Replaced the use of synchronized with an ExecutorService for more efficient thread management.
- Reused the same thread pool for multiple metric collection tasks.
- General improvements in code organization and exception handling.
- Optimized parallel metric collection to accelerate the revision analysis process.
- Added explicit disposal of unused objects to free up memory as needed.
- Other minor code improvements and fixes to enhance efficiency and robustness.

These changes significantly enhance the performance and stability of Git project analysis.
